zebra-chain/src/chain_tip.rs

```diff
@@ -95,26 +95,29 @@ pub trait ChainTip {
         Some(estimator.estimate_height_at(now))
     }
 
-    /// Return an estimate of how many blocks there are ahead of Zebra's best chain tip
-    /// until the network chain tip, and Zebra's best chain tip height.
+    /// Return an estimate of how many blocks there are ahead of Zebra's best chain tip until the
+    /// network chain tip, and Zebra's best chain tip height.
+    ///
+    /// The first element in the returned tuple is the estimate.
+    /// The second element in the returned tuple is the current best chain tip.
     ///
     /// The estimate is calculated based on the current local time, the block time of the best tip
     /// and the height of the best tip.
     ///
-    /// This estimate may be negative if the current local time is behind the chain tip block's timestamp.
+    /// This estimate may be [`None`] if the current local time is behind the chain tip block's
+    /// timestamp.
     fn estimate_distance_to_network_chain_tip(
         &self,
         network: Network,
-    ) -> Option<(i32, block::Height)> {
+    ) -> Option<(block::Height, block::Height)> {
         let (current_height, current_block_time) = self.best_tip_height_and_block_time()?;
 
         let estimator =
             NetworkChainTipHeightEstimator::new(current_block_time, current_height, network);
 
-        Some((
-            estimator.estimate_height_at(Utc::now()) - current_height,
-            current_height,
-        ))
+        let distance_to_tip = estimator.estimate_height_at(Utc::now()) - current_height;
+
+        distance_to_tip.map(|distance_to_tip| (distance_to_tip, current_height))
```

Comment:
In some cases, the callers of this function use the tip height, even if the estimate is below the current tip. It's not clear to me if this PR keeps that requirement, or breaks it.

For example:
* the `getblocktemplate` RPC could return "No Chain tip available yet" if the estimate is below the current tip, but it should instead return a template
* the `getblockchaininfo.estimatedheight` field should return the current best height, rather than a "No Chain tip available yet" error for the whole RPC, see https://zcash.github.io/rpc/getblockchaininfo.html

Are there existing unit tests for these two edge cases, or should we add them to this PR?

Can we return `(Option<Height>, Option<Height>)` here? Or a struct with named `Option<Height>` fields?
Now that the two tuple fields are the same type, it's easy for the caller to mix them up. And since the fields aren't named, reviewers will find it hard to pick up a mistake like that.

Reply:
I thought about this a bit more, and I think we're actually trying to represent four different outcomes here:
- Empty: the blockchain is empty
- Syncing: the `estimated_height` is above the `state_height`
- SyncedEqual: the estimated and `state_height` are equal
- SyncedAbove: the estimated height is above the `state_height`

Would it help to have an enum, or is that too complicated?

If we don't have any code that cares about the difference between the last two outcomes, we could represent those two outcomes in the same way, with just one enum variant and height field.

Reply:
> * the `getblocktemplate` RPC could return "No Chain tip available yet" if the estimate is below the current tip, but it should instead return a template

This method would've panicked before this PR if it tried to subtract a bigger height from the estimated network height.

I still think a minimum of zero (i.e. `Some((distance_to_tip.unwrap_or(block::Height(0)), current_height))`) would work nicely here.

Reply:
> I can manually test both mining and a full sync before a merge.

> Are there existing unit tests for these two edge cases, or should we add them to this PR?

I don't think we have them. I'll try to add them.

> Would it help to have an enum, or is that too complicated?

The enum you outlined sounds good.

Reply:
> > * the `getblocktemplate` RPC could return "No Chain tip available yet" if the estimate is below the current tip, but it should instead return a template
> 
> This method would've panicked before this PR if it tried to subtract a bigger height from the estimated network height.

I'm not sure where the panic is, can you explain it to me?

I think I've seen the estimate below the tip before, but I'm not sure.

> I still think a minimum of zero (i.e. `Some((distance_to_tip.unwrap_or(block::Height(0)), current_height))`) would work nicely here.

This code is not consensus-critical, so it doesn't really matter what we do here. But is there any RPC specification, RPC error, or any other code that cares about the difference between a negative and a zero distance to the tip?

Reply:
> I'm not sure where the panic is, can you explain it to me?

It *was* [here](https://github.com/ZcashFoundation/zebra/pull/6330/files#diff-929cb1ee6ff0e87f011f57a1a5f68132642202776e071648e4138ca28ce032d2L96-L97), where a `Height` was being subtracted from a possibly lower `Height` in this `estimate_distance_to_network_chain_tip()` method.

Reply:
> Are there existing unit tests for these two edge cases, or should we add them to this PR?

There aren't any existing unit tests that I'm aware of, but I think we should add those in a different PR.